x/perpetual/keeper/pool.go

```diff
@@ -68,3 +79,238 @@ func (k Keeper) RemoveLegacyPool(ctx sdk.Context, index uint64) {
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		store.Delete(iterator.Key())
+	}
+}
+
+func (k Keeper) GetAllBorrowRate(ctx sdk.Context) []types.InterestBlock {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.InterestRatePrefix)
+	iterator := sdk.KVStorePrefixIterator(store, nil)
+	defer iterator.Close()
+
+	interests := []types.InterestBlock{}
+	for ; iterator.Valid(); iterator.Next() {
+		interest := types.InterestBlock{}
+		k.cdc.MustUnmarshal(iterator.Value(), &interest)
+
+		interests = append(interests, interest)
+	}
+	return interests
+}
+
+func (k Keeper) GetBorrowRate(ctx sdk.Context, startBlock uint64, pool uint64, borrowed sdk.Dec) sdk.Dec {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.InterestRatePrefix)
+	currentBlockKey := types.GetInterestRateKey(uint64(ctx.BlockHeight()), pool)
+	startBlockKey := types.GetInterestRateKey(startBlock, pool)
+
+	// note: exclude start block
+	if store.Has(startBlockKey) && store.Has(currentBlockKey) && startBlock != uint64(ctx.BlockHeight()) {
+		bz := store.Get(startBlockKey)
+		startInterestBlock := types.InterestBlock{}
+		k.cdc.MustUnmarshal(bz, &startInterestBlock)
+
+		bz = store.Get(currentBlockKey)
+		endInterestBlock := types.InterestBlock{}
+		k.cdc.MustUnmarshal(bz, &endInterestBlock)
+
+		totalInterest := endInterestBlock.InterestRate.Sub(startInterestBlock.InterestRate)
+
+		newInterest := borrowed.Mul(totalInterest)
+
+		return newInterest
+	}
+
+	if !store.Has(startBlockKey) && store.Has(currentBlockKey) {
+		iterator := sdk.KVStorePrefixIterator(store, nil)
+		defer iterator.Close()
+
+		firstStoredBlock := uint64(0)
+		if iterator.Valid() {
+			interestBlock := types.InterestBlock{}
+			firstStoredBlock = sdk.BigEndianToUint64(iterator.Key())
+			k.cdc.MustUnmarshal(iterator.Value(), &interestBlock)
+		}
+		if firstStoredBlock > startBlock {
+			bz := store.Get(currentBlockKey)
+			endInterestBlock := types.InterestBlock{}
+			k.cdc.MustUnmarshal(bz, &endInterestBlock)
+
+			totalInterest := endInterestBlock.InterestRate
+			newInterest := borrowed.Mul(totalInterest)
+			return newInterest
+		}
+	}
+	params, found := k.GetPool(ctx, pool)
+	if !found {
+		return sdk.ZeroDec()
+	}
+	newInterest := borrowed.Mul(params.BorrowInterestRate)
+	return newInterest
+}
+
+func (k Keeper) SetFundingRate(ctx sdk.Context, block uint64, pool uint64, funding types.FundingRateBlock) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.FundingRatePrefix)
+	prev := types.GetFundingRateKey(block-1, pool)
+	key := types.GetFundingRateKey(block, pool)
+	if store.Has(prev) {
+		lastBlock := types.FundingRateBlock{}
+		bz := store.Get(prev)
+		k.cdc.MustUnmarshal(bz, &lastBlock)
+		funding.FundingRate = funding.FundingRate.Add(lastBlock.FundingRate)
+		funding.TotalCustodyLong = funding.TotalCustodyLong.Add(lastBlock.TotalCustodyLong)
+		funding.TotalCustodyShort = funding.TotalCustodyShort.Add(lastBlock.TotalCustodyShort)
```

Comment:
@amityadav0 why are we adding up block-1 and block numbers, each block should only have knowledge of the funding rate and the long and short custody value as from one block to another you can have an increased on decreased custody amount depending on perpetual activities.

Reply:
this was done similar to interest rate, as interest rate accumulates over time, and if we want to get interest rate between two blocks, we can get that by subtracting previous block
example:
Interest rate 1 1 2 1 1
Store.            1 2 4 5 6
Now to get interest rate between 2-4
we can do position(4) - position(1) = 4